jerry-core/ecma/builtin-objects/ecma-builtin-helpers.c

```diff
@@ -140,58 +140,48 @@ ecma_builtin_helper_object_to_string (const ecma_value_t this_arg) /**< this arg
 
     ecma_object_t *obj_p = ecma_get_object_from_value (obj_this);
 
-    type_string = ecma_object_get_class_name (obj_p);
+    builtin_tag = ecma_object_get_class_name (obj_p);
 
 #if ENABLED (JERRY_ES2015)
-    ecma_value_t tag_value = ecma_op_object_get_by_symbol_id (obj_p, LIT_GLOBAL_SYMBOL_TO_STRING_TAG);
+    ecma_value_t is_array = ecma_is_value_array (obj_this);
 
-    if (ECMA_IS_VALUE_ERROR (tag_value))
+    if (ECMA_IS_VALUE_ERROR (is_array))
     {
       ecma_deref_object (obj_p);
-      return tag_value;
+      return is_array;
     }
 
-    if (ecma_is_value_string (tag_value))
+    if (ecma_is_value_true (is_array))
+    {
+      builtin_tag = LIT_MAGIC_STRING_ARRAY_UL;
+    }
+
+    ecma_value_t tag = ecma_op_object_get_by_symbol_id (obj_p, LIT_GLOBAL_SYMBOL_TO_STRING_TAG);
+
+    if (ECMA_IS_VALUE_ERROR (tag))
+    {
+      ecma_deref_object (obj_p);
+      return tag;
+    }
+
+    if (ecma_is_value_string (tag))
     {
       ecma_deref_object (obj_p);
-      return ecma_builtin_helper_object_to_string_tag_helper (tag_value);
+      return ecma_builtin_helper_object_to_string_tag_helper (tag);
     }
 
-    ecma_free_value (tag_value);
+    ecma_free_value (tag);
 #endif /* ENABLED (JERRY_ES2015) */
 
     ecma_deref_object (obj_p);
   }
 
-  ecma_string_t *ret_string_p;
-
-  /* Building string "[object #type#]" where type is 'Undefined',
-     'Null' or one of possible object's classes.
-     The string with null character is maximum 27 characters long. */
-  const lit_utf8_size_t buffer_size = 27;
-  JERRY_VLA (lit_utf8_byte_t, str_buffer, buffer_size);
-
-  lit_utf8_byte_t *buffer_ptr = str_buffer;
-
-  const lit_magic_string_id_t magic_string_ids[] =
-  {
-    LIT_MAGIC_STRING_LEFT_SQUARE_CHAR,
-    LIT_MAGIC_STRING_OBJECT,
-    LIT_MAGIC_STRING_SPACE_CHAR,
-    type_string,
-    LIT_MAGIC_STRING_RIGHT_SQUARE_CHAR
-  };
-
-  for (uint32_t i = 0; i < sizeof (magic_string_ids) / sizeof (lit_magic_string_id_t); ++i)
-  {
-    buffer_ptr = lit_copy_magic_string_to_buffer (magic_string_ids[i], buffer_ptr,
-                                                  (lit_utf8_size_t) ((str_buffer + buffer_size) - buffer_ptr));
-    JERRY_ASSERT (buffer_ptr <= str_buffer + buffer_size);
-  }
+  ecma_stringbuilder_t builder = ecma_stringbuilder_create_raw ((lit_utf8_byte_t *) "[object ", 8);
```

Comment:
What about creating a magic string for `[object `? If this string is in the binary file, the JS code should be able to take advantage of it.

Reply:
I don't think there's a lot to take advantage of. This string practically never happens by itself, so the engine wouldn't be able to use it for anything. It would just clutter the magic strings.

Reply:
My reasoning is the string is already in the binary, and if a JS will ever use it: it is a gain, if never: no disadvantage. Magic string search is logaritmic, you need to double them to add one new step.